Fix endian-related errors in frontend.

redstar opened the pull request:
There are still some places where the frontend is not endian-neutral.
I identified these places while porting LDC to big endian PowerPC.
I have no good idea how to avoid the preprocessor here but I am open
to suggestions.

The code checks for **BIG_ENDIAN** which is not defined by default
and therefore no code paths are changed. On a big endian platform this
has to be defined separately (there is no predefined compiler macro
you could use).

yebblies commented:
I would prefer these were written as endian-independent loads and shifts instead of `#if`ed

redstar commented:
Any idea how to do it? The variable `v` always contains the value to copy. Depending on the endian type the first or the last `sz` bytes needs to be copied.

redstar commented:
A possible solution would be a `switch`, e.g.:

            	  switch (sz)
            	  {
            	      case 1: *((unsigned char *) s) = (unsigned char) v; break;
            	      case 2: *((unsigned short *) s) = (unsigned short) v; break;
            	      case 4: *((unsigned int *) s) = (unsigned int) v; break;
            	      case 8: *((unsigned long *) s) = (unsigned long) v; break;
            	  }

yebblies commented:
Yes, that switch looks right.  Although I think we have fixed-width integral types defined somewhere.

ibuclaw commented:
> Yes, that switch looks right. Although I think we have fixed-width integral types defined somewhere.

Probably in globals.h.  IIRC, it's ``d_uns8``, ``d_uns16``, ``d_uns32``, and ``d_uns64``.

redstar commented:
I replaced the `#if` with the switch statement.

ibuclaw commented:
LTGM

yebblies commented:
I assume this is failing because you're not offsetting the pointer with the loop index any more.  Could you put this in a new Port function somewhere rather then copy-pasting?

ibuclaw commented:
> I assume this is failing because you're not offsetting the pointer with the loop index any more.

Ouch, just noticed that too

redstar commented:
Any hint why magicport fails?

yebblies commented:
`constfold.d(1709): Error: no property 'valcpy' for type 'Port'`

You'll need to hand-port `valcpy` to `port.d`.

yebblies commented:
Conflict...

yebblies commented:
Ping.  This should be fairly easy to port over to ddmd.

redstar committed:
Fix endian-related errors in frontend.

There are still some places where the frontend is not endian-neutral.
I identified these places while porting LDC to big endian PowerPC.

A new function valcpy() is added to ddmd.root.port to copy values
regardless of endianess.

redstar commented:
Finally updated for ddmd.

yebblies commented on src/root/port.h:
```diff
@@ -75,6 +75,7 @@ struct Port
     static unsigned readlongBE(void* buffer);
     static unsigned readwordLE(void* buffer);
     static unsigned readwordBE(void* buffer);
+    static void valcpy(void *dst, const uint64_t val, const size_t size);
```
Signature doesn't match?

yebblies commented on src/root/port.d:
```diff
@@ -303,4 +303,16 @@ extern (C++) struct Port
+    {
+        switch (size)
+        {
+            case 1: *cast(ubyte *)dst = cast(ubyte) val; break;
```
DDMD style is no space after cast

yebblies commented on src/constfold.d:
```diff
@@ -1454,7 +1454,7 @@ extern (C++) UnionExp Cat(Type type, Expression e1, Expression e2)
             void* s = mem.xmalloc(len * sz);
             if (t.ty == tn.ty)
-                memcpy(s, &v, sz);
+                valcpy(s, v, sz);
```
Should be `Port.valcpy`?

redstar commented:
Fixed.

yebblies commented:
Auto-merge toggled on
